### PR TITLE
sprintf -> snprintf

### PR DESCRIPTION
**Pull request by seanm**
The last commit needs discussion of how to deprecate an API and replace it with another, and to choose its name...

### CODE REVIEW

**Comment by seanm** [comment]
> I would suggest breaking it up to do one directory at a time, E,g, for libdap4, libhdf5, etc.

Breaking what up?  The big commit?  By my count that would be 17 commits.

Is this to make bisecting possible regressions easier?  (Trying to understand why...)

File: h5_test/tst_h_compounds.c [hunk 1]
```diff
@@ -452,7 +452,7 @@ main()
          hr_data_out[i].starfleet_id = i;
          hr_data_out[i].svc_rec.i1 = 95;
          hr_data_out[i].svc_rec.i2 = 90;
-         if (sprintf(hr_data_out[i].name, "alien_%d", i) < 0) ERR;
+         if (snprintf(hr_data_out[i].name, sizeof(hr_data_out[i].name), "alien_%d", i) < 0) ERR;
```
Comment (edwardhartnett) [inline h5_test/tst_h_compounds.c:455]: It's not clear to be why this change is an improvement over the original.

File: libdap2/dapodom.c [hunk 2]
```diff
@@ -81,7 +81,7 @@ dapodom_print(Dapodometer* odom)
     if(odom->rank == 0) {
 	strlcat(line,"[]",sizeof(line));
     } else for(i=0;i<odom->rank;i++) {
-	sprintf(tmp,"[%lu/%lu:%lu:%lu]",
+	snprintf(tmp,sizeof(tmp),"[%lu/%lu:%lu:%lu]",
```
Comment (edwardhartnett) [inline libdap2/dapodom.c:84]: It seems like the only reason to use snprintf() is in case of an incorrectly terminated string. In that case, it will prevent an overflow.

But not if you just use sizeof() on the original string. In that case, you will still get the exact same overflow. Or am I missing something?

Comment (seanm) [reply libdap2/dapodom.c:84]: Let me flip it around: what is the reason to use `sprintf` instead of `snprintf` ?

File: h5_test/tst_h_compounds.c (hunk 1 above)
Comment (edwardhartnett) [reply h5_test/tst_h_compounds.c:455]: OK, I guess that's a good point.

File: libdap2/dapodom.c (hunk 2 above)
Comment (edwardhartnett) [reply libdap2/dapodom.c:84]: OK, now I get it.

**Comment by WardF** [comment]
Agreed this is a lot of files, but given the consistent nature of the change, it is something I can commit to spending some time going through.  The main issue will be if there are any strings which should not be `NULL` terminated but now are.  I can't think of any instance off the top of my head, but it's still part of the job XD.  Also, the `Appveyor` CI indicates a failure compiling w/ Visual Studio; I haven't looked at that more closely, but we'd want to fix that too.

**Comment by seanm** [comment]
Also the last commit is a WIP... A public API needs to change to be given the length of a buffer... I used a stupid placeholder name for the new function... What would you like to call it?

File: libhdf5/nc4hdf.c [hunk 3]
```diff
@@ -708,13 +708,13 @@ write_quantize_att(NC_VAR_INFO_T *var)
     switch (var->quantize_mode)
     {
 	case NC_QUANTIZE_BITGROOM:
-	    sprintf(att_name, "%s", NC_QUANTIZE_BITGROOM_ATT_NAME);
+	    snprintf(att_name, sizeof(att_name), "%s", NC_QUANTIZE_BITGROOM_ATT_NAME);
```
[the remark this answers is not in the document]
Comment (seanm) [reply libhdf5/nc4hdf.c:711]: Here too the idea is simply that it's never preferable to use `sprintf` over `snprintf`.

**Comment by seanm** [comment]
@WardF @edwardhartnett @DennisHeimbigner I'd like to fix up the last of the commits in this PR...

How does NetCDF deal with deprecating and replacing APIs?  i.e. for this part:

```diff
-extern void cdRel2Iso(cdCalenType timetype, char* relunits, int separator, double reltime, char* chartime);
+extern void cdRel2IsoNew(cdCalenType timetype, char* relunits, int separator, double reltime, char* chartime, size_t chartime_size);
+extern void cdRel2Iso(cdCalenType timetype, char* relunits, int separator, double reltime, char* chartime); // deprecated
```

1. what shall we call the new API?
2. do you have a deprecation macro that does similar to [C23's `[[deprecated]]`](https://en.cppreference.com/w/c/language/attributes/deprecated)?

Thanks.

**Comment by seanm** [comment]
In the hopes of actually getting this merged, I moved the WIP part to a new PR: https://github.com/Unidata/netcdf-c/pull/2743

This one is now ready to merge, after review.

**Comment by seanm** [comment]
Merged conflicts fixed.

@WardF @DennisHeimbigner this one is ready to merge IMHO, pending your reviews.

**Review by WardF — APPROVED** [review]

**Comment by WardF** [comment]
Testing on Windows for completeness sake; our current CI builds but doesn't run the tests, due to hangs on the underlying systems.  It's been a while, I need to double-check that again, we really need to get Windows CI running again (not just building).

**Comment by seanm** [comment]
@WardF did your Windows test pass?

@edwardhartnett are you good with these changes now?
